anchore_engine/services/policy_engine/engine/loaders.py

```diff
@@ -1177,31 +1183,69 @@ def load_cpes_from_syft_output_with_fallback(self, analysis_json, image):
+                )
+                cpes.extend(os_pkgs_cpes)
+
         return cpes
 
     def extract_syft_cpes(self, allcpes, package_dict, image, pkg_type):
         cpes = []
         for pkg_key, pkg_json_str in package_dict.items():
             pkg = safe_extract_json_value(pkg_json_str)
             pkg_cpes = pkg.get("cpes", [])
-            for cpe in pkg_cpes:
-                decomposed_cpe = self.decompose_cpe(cpe)
-                cpekey = ":".join(decomposed_cpe + [pkg_key])
-
-                if cpekey not in allcpes:
-                    allcpes[cpekey] = True
-                    image_cpe = ImageCpe()
-                    image_cpe.pkg_type = pkg_type
-                    image_cpe.pkg_path = pkg_key
-                    image_cpe.cpetype = decomposed_cpe[2]
-                    image_cpe.vendor = decomposed_cpe[3]
-                    image_cpe.name = decomposed_cpe[4]
-                    image_cpe.version = decomposed_cpe[5]
-                    image_cpe.update = decomposed_cpe[6]
-                    image_cpe.meta = decomposed_cpe[7]
-                    image_cpe.image_user_id = image.user_id
-                    image_cpe.image_id = image.id
-                    cpes.append(image_cpe)
+
+            cpes.extend(
+                self._make_image_cpes(image, allcpes, pkg_cpes, pkg_type, pkg_key)
+            )
+
+        return cpes
+
+    def _make_image_cpes(self, image, allcpes, pkg_cpes, pkg_type, pkg_path):
+        cpes = []
+        for cpe in pkg_cpes:
+            decomposed_cpe = self.decompose_cpe(cpe)
+            cpekey = ":".join(decomposed_cpe + [pkg_path])
+
+            if cpekey not in allcpes:
+                allcpes[cpekey] = True
+                image_cpe = ImageCpe()
+                image_cpe.pkg_type = pkg_type
+                image_cpe.pkg_path = pkg_path
+                image_cpe.cpetype = decomposed_cpe[2]
+                image_cpe.vendor = decomposed_cpe[3]
+                image_cpe.name = decomposed_cpe[4]
+                image_cpe.version = decomposed_cpe[5]
+                image_cpe.update = decomposed_cpe[6]
+                image_cpe.meta = decomposed_cpe[7]
+                image_cpe.image_user_id = image.user_id
+                image_cpe.image_id = image.id
+                cpes.append(image_cpe)
+
+        return cpes
+
+    def extract_syft_cpes_for_os_packages(self, allcpes, package_dict, image):
+        """
+        Utility function for parsing cpes for os packages only. Mostly a duplicate of extract_syft_cpes with the exception of pkg_type and pkg_path.
+        For os packages pkg_path is always pkgdb. pkg_type is picked up from the data instead of passing it as a function argument
+
+        """
+        cpes = []
+        for pkg_key, pkg_json_str in package_dict.items():
+            pkg = safe_extract_json_value(pkg_json_str)
+            pkg_path = f"pkgdb/{pkg_key}"  # os package location is always pkgdb. this is a deviation to tie a package with it's cpes
```

Comment:
Is there ever a scenario where `pkg_key` could be None?

Reply:
That should never be the case, pkg_key and pkg_json_str are the key-value pairs in a dictionary. Granted that doesn't prevent it from a null pointer reference, but worst case scenario only one such reference will exist in the dictionary. And even in that case, the logic won't bail